ndk/src/midi.rs

```diff
@@ -0,0 +1,292 @@
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn ptr(&self) -> NonNull<ffi::AMidiInputPort> {
+        self.ptr
+    }
+
+    /// Sends data to this port.
+    pub fn send(&self, buffer: &[u8]) -> Result<usize> {
+        let num_bytes_sent =
+            unsafe { ffi::AMidiInputPort_send(self.ptr.as_ptr(), buffer.as_ptr(), buffer.len()) };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(MediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int)).unwrap_err())
+        }
+    }
+
+    /// Sends a message with a 'MIDI flush command code' to this port.
+    ///
+    /// This should cause a receiver to discard any pending MIDI data it may have accumulated and
+    /// not processed.
+    pub fn send_flush(&self) -> Result<()> {
+        let result = unsafe { ffi::AMidiInputPort_sendFlush(self.ptr.as_ptr()) };
+        MediaError::from_status(result)
+    }
+
+    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_sendWithTimestamp(
+                self.ptr.as_ptr(),
+                buffer.as_ptr(),
+                buffer.len(),
+                timestamp,
+            )
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(MediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int)).unwrap_err())
+        }
+    }
+}
+
+impl<'a> Drop for MidiInputPort<'a> {
+    fn drop(&mut self) {
+        unsafe { ffi::AMidiInputPort_close(self.ptr.as_ptr()) };
+    }
+}
+
+pub struct MidiOutputPort<'a> {
+    ptr: NonNull<ffi::AMidiOutputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiOutputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiOutputPort")
+            .field("inner", &self.ptr)
+            .finish()
+    }
+}
+
+impl<'a> MidiOutputPort<'a> {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::AMidiOutputPort`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::AMidiOutputPort>) -> Self {
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn ptr(&self) -> NonNull<ffi::AMidiOutputPort> {
+        self.ptr
+    }
+
+    /// Receives the next pending MIDI message.
+    ///
+    /// To retrieve all pending messages, the client should repeatedly call this method until it
+    /// returns [`Ok(MidiOpcode::NoMessage)`].
+    ///
+    /// Note that this is a non-blocking call. If there are no Midi messages are available, the
+    /// function returns [`Ok(MidiOpcode::NoMessage)`] immediately (for 0 messages received).
+    pub fn receive(&self, buffer: &mut [u8]) -> Result<MidiOpcode> {
```

Comment:
Maybe add an extra paragraph explaining that the `MidiOpcode` contains the `length` which the user must use to bound their slice?

Alternatively I've been toying with `buffer: &mut &mut [u8]`-like APIs, where I update the caller slice to `*buffer = &mut buffer[..num_bytes_received]` so that the user only gets the valid number of bytes back. This becomes more important when you start supporting `[MaybeUninit<u8>]` inputs (which _may_ be useful here!). Alternatively `MidiOpcode` could borrow and return the valid sub-slice (instead of `length`?) so that the user has direct access to these.

Unfortunately all slice helpers around `MaybeUninit` are eternally stuck in "nightly-only experimental" land: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_assume_init_mut

I'll leave this choice up to you, depending on convenience. The user quite likely already has something along the form of (i.e. it is quite hard to ignore/overlook `MidiOpcode::Data::length`):

```rust
match receive(&mut buffer)? {
    NoMessage => break,
    Data { length, timestamp: _ } => {
        let useful = &buffer[..length as usize];
        // ...
    }
    Flush => ...,
}
```